File: corelibrary/src/main/java/pers/cxd/corelibrary/util/reflection/FieldKey.java
package pers.cxd.corelibrary.util.reflection;

import java.util.Objects;

public class FieldKey {

    Class<?> clazz;
    String filedName;
    private boolean inUse;
    private FieldKey next;

    private FieldKey(Class<?> clazz, String filedName) {
        this.clazz = clazz;
        this.filedName = filedName;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        FieldKey fieldKey = (FieldKey) o;
        return Objects.equals(clazz, fieldKey.clazz) &&
                Objects.equals(filedName, fieldKey.filedName);
    }

    @Override
    public int hashCode() {
        return Objects.hash(clazz, filedName);
    }

    public void markInUse() {
        inUse = true;
    }

    private static FieldKey sPool;
    private static int sPoolSize;
    private static final int MAX_POOL_SIZE = 10;

    public static FieldKey obtain(Class<?> clazz, String filedName){
        synchronized (FieldKey.class) {
            if (sPool != null) {
                FieldKey fk = sPool;
                sPool = fk.next;
                fk.next = null;
                sPoolSize--;
                fk.clazz = clazz;
                fk.filedName = filedName;
                return fk;
            }
        }
        return new FieldKey(clazz, filedName);
    }

    public void recycle(){
        if (inUse){
            throw new IllegalStateException(this + " is in use, can't recycle");
        }
        clazz = null;
        filedName = null;
        synchronized (FieldKey.class) {
            if (sPoolSize < MAX_POOL_SIZE) {
                next = sPool;
                sPool = this;
                sPoolSize++;
            }
        }
    }

}

File: corelibrary/src/main/java/pers/cxd/corelibrary/util/reflection/FieldKey.kt
package pers.cxd.corelibrary.util.reflection

import java.util.*

class FieldKey private constructor(var clazz: Class<*>?, var filedName: String?) {
    private var inUse = false
    private var next: FieldKey? = null
    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val fieldKey = o as FieldKey
        return clazz == fieldKey.clazz &&
                filedName == fieldKey.filedName
    }

    override fun hashCode(): Int {
        return Objects.hash(clazz, filedName)
    }

    fun markInUse() {
        inUse = true
    }

    fun recycle() {
        check(!inUse) { "$this is in use, can't recycle" }
        clazz = null
        filedName = null
        synchronized(FieldKey::class.java) {
            if (sPoolSize < MAX_POOL_SIZE) {
                next = sPool
                sPool = this
                sPoolSize++
            }
        }
    }

    companion object {
        private const val MAX_POOL_SIZE = 10
        private var sPool: FieldKey? = null
        private var sPoolSize = 0
        fun obtain(clazz: Class<*>?, filedName: String?): FieldKey {
            synchronized(FieldKey::class.java) {
                if (sPool != null) {
                    val fk = sPool
                    sPool = fk!!.next
                    fk.next = null
                    sPoolSize--
                    fk.clazz = clazz
                    fk.filedName = filedName
                    return fk
                }
            }
            return FieldKey(clazz, filedName)
        }
    }
}